## PROJETO DE SISTEMAS LINEARES

*Projeto desenvolvido pela aluna Ana Beatriz de Souza Nogueira para a disciplina de Métodos Numéricos referente ao conteúdo de resoluções de Sistemas Lineares.*

# ***Dieta Alimentar Equilibrada***
## *PROBLEMA*

Uma jovem estudante de Engenharia Elétrica do IFPB está fazendo seus projetos e se sente muito cansada. Decide então pesquisar no Google (erradíssimo) uma dieta com quilocalorias, lipídeos e proteínas suficientes para as atividades acadêmicas, afinal, a graduação consome todas as energias existentes no seu corpo.

A dieta diz que a estudante precisa consumir diariamente 2072Kcal (quilocalorias), 220g de proteínas e 19,3g de lipídeos. Dito isto, ela teve a brilhante ideia de fazer essa dieta a base dos alimentos preferidos dela: lasanha de frango, risoto de limão e temaki.

Na tabela abaixo contém a quantidade das substâncias encontradas em cada alimento.

![alt text](https://i.ibb.co/NCh0DMg/tabelinha.png)

Para a estudante saber quantos gramas de cada um destes alimentos ela deve ingerir diariamente, montou o sistema abaixo, Sendo Kcal (quilocalorias), prot (proteínas) e lip (lipídeos).

$$ kcal(lasanha) + kcal(risoto) + kcal(temaki) = kcal(total) $$
$$ prot(lasanha) + prot(risoto) + prot(temaki) = prot(total) $$
$$ lip(lasanha) + lip(risoto) + lip(temaki) = lip(total) $$

Tendo isso em vista, ela substituiu os valores contidos na tabela no sistema, ficando assim, sendo l (lasanha), r (risoto) e t (temaki):

$$ 128(l) + 159(r) + 63(t) = 2072 $$
$$ 2.5(l) + 32(r) + 0.2(t) = 220 $$
$$ 0.2(l) + 2.5(r) + 4.2(t) = 19.3 $$

Logo, para saber quantas porções de 100g de cada alimento ela deve comer, basta resolver o sistema. E como ela estudou em Métodos Numéricos como resolver sistemas lineares, essa tarefa será izi pizi heheheh.




## ***Resolvendo o Sistema utilizando Gauss Ingênuo***


A resolução de sistemas lineares utilizando o método de Gauss Ingênuo é bem simples e intuitiva.


In [ ]:
#Incluindo bibliotecas

import numpy as np
np.set_printoptions(formatter={'float': lambda x: "{:.2f}".format(x)})

In [ ]:
#Matriz Gauss Ingênuo

m_gi = np.array([[128, 159, 63, 2072], [2.5, 32, 0.2, 220], [0.2, 2.5, 4.2, 19.3]])

In [ ]:
#Primeiro pivô

pivo1_gi = m_gi[0,0]
pivo1_gi

128.0

In [ ]:
#Primeiro multiplicador

m1_gi = m_gi[1,0]/pivo1_gi
m1_gi

0.01953125

In [ ]:
#Zerando o primeiro elemento abaixo do pivô

m_gi[1,:] = m_gi[1,:] - m_gi[0,:]*m1_gi
m_gi[1,:]

array([0.00, 28.89, -1.03, 179.53])

In [ ]:
#Segundo multiplicador

m2_gi = m_gi[2,0]/pivo1_gi
m2_gi

0.0015625

In [ ]:
#Zerando o segundo elemento abaixo do pivô

m_gi[2,:] = m_gi[2,:] - m_gi[0,:]*m2_gi
m_gi[2,:]

array([0.00, 2.25, 4.10, 16.06])

In [ ]:
m_gi

array([[128.00, 159.00, 63.00, 2072.00],
       [0.00, 28.89, -1.03, 179.53],
       [0.00, 2.25, 4.10, 16.06]])

In [ ]:
#Segundo pivô

pivo2_gi = m_gi[1,1]
pivo2_gi

28.89453125

In [ ]:
#Terceiro multiplicador

m3_gi = m_gi[2,1]/pivo2_gi
m3_gi

0.07792348249290253

In [ ]:
#Zerando o primeiro elemento abaixo do segundo pivô

m_gi[2,:] = m_gi[2,:] - m_gi[1,:]*m3_gi
m_gi[2,:]

array([0.00, 0.00, 4.18, 2.07])

In [ ]:
m_gi

array([[128.00, 159.00, 63.00, 2072.00],
       [0.00, 28.89, -1.03, 179.53],
       [0.00, 0.00, 4.18, 2.07]])

In [ ]:
#Porções de cada alimento
p_gi = np.zeros(3)

In [ ]:
p_gi[2] = m_gi[2,3]/m_gi[2,2]

In [ ]:
p_gi[1] = (m_gi[1,3] - m_gi[1,2]*p_gi[2])/m_gi[1,1]

In [ ]:
p_gi[0] = (m_gi[0,3] - m_gi[0,2]*p_gi[2] - m_gi[0,1]*p_gi[1])/m_gi[0,0]

In [ ]:
p_gi

array([8.20, 6.23, 0.50])

## ***Resolvendo o Sistema utilizando Gauss com Pivotação Parcial***

O método de Gauss com pivotação parcial é utilizado quando o elemento [0,0] é muito próximo ou igual a zero. Basta trocar a linha do maior elemento da primeira coluna com a linha 0, evitando assim, divisões por 0.

In [ ]:
#Matriz Gauss com Pivotação Parcial

m_gp = np.array([[128, 159, 63, 2072], [2.5, 32, 0.2, 220], [0.2, 2.5, 4.2, 19.3]])

In [ ]:
#Indice com maior valor na coluna 0

np.argmax(np.abs(m_gp[:,0]))

0

O índice que tem maior valor na primeira coluna é o índice 0, não será necessário utilizar o método da pivotação parcial, logo, a resolução será idêntica ao método de Gauss Ingênuo.

In [ ]:
#Primeiro pivô

pivo1_gp = m_gp[0,0]
pivo1_gp

128.0

In [ ]:
#Primeiro multiplicador

m1_gp = m_gp[1,0]/pivo1_gp
m1_gp

0.01953125

In [ ]:
#Zerando o elemento abaixo do pivô

m_gp[1,:] = m_gp[1,:] - m_gp[0,:]*m1_gp
m_gp[1,:]

array([0.00, 28.89, -1.03, 179.53])

In [ ]:

#Segundo multiplicador

m2_gp = m_gp[2,0]/pivo1_gp
m2_gp

0.0015625

In [ ]:
#Zerando o segundo elemneto abaixo do pivô

m_gp[2,:] = m_gp[2,:] - m_gp[0,:]*m2_gp
m_gp[2,:]

array([0.00, 2.25, 4.10, 16.06])

In [ ]:
m_gp

array([[128.00, 159.00, 63.00, 2072.00],
       [0.00, 28.89, -1.03, 179.53],
       [0.00, 2.25, 4.10, 16.06]])

In [ ]:
#Segundo pivô

pivo2_gp = m_gp[1,1]
pivo2_gp

28.89453125

In [ ]:
#Terceito multiplicador

m3_gp = m_gp[2,1]/pivo2_gp
m3_gp

0.07792348249290253

In [ ]:
#Zerando o elemento abaixo do segundo pivô

m_gp[2,:] = m_gp[2,:] - m_gp[1,:]*m3_gp
m_gp[2,:]

array([0.00, 0.00, 4.18, 2.07])

In [ ]:
m_gp

array([[128.00, 159.00, 63.00, 2072.00],
       [0.00, 28.89, -1.03, 179.53],
       [0.00, 0.00, 4.18, 2.07]])

In [ ]:
#Porções de cada alimento

p_gp = np.zeros(3)

In [ ]:
p_gp[2] = m_gp[2,3]/m_gp[2,2]

In [ ]:
p_gp[1] = (m_gp[1,3] - m_gp[1,2]*p_gp[2])/m_gp[1,1]

In [ ]:
p_gp[0] = (m_gp[0,3] - m_gp[0,2]*p_gp[2] - m_gp[0,1]*p_gp[1])/m_gp[0,0]

In [ ]:
p_gp

array([8.20, 6.23, 0.50])

## ***Resolvendo o Sistema utilizando Gauss Seidel***




O método de Gauss Seidel é um método iterativo para a resolução de sistemas lineares. Esse método, SEM DÚVIDAS, foi o método que a estudante mais gostou.

In [ ]:
max_it = 100 #Número de iterações máximas
tolerancia = 0.0001 #Tolerância do erro
p_gs = np.zeros(3) #Começando todos os valores das porções sendo 0

for i in range(max_it):
    p_gs_anterior = p_gs.copy() #Representando cada coeficiente em função dos outros
    p_gs[0] = (2072 - 159*p_gs[1] - 63*p_gs[2])/128
    p_gs[1] = (220 - 2.5*p_gs[0] - 0.2*p_gs[2])/32
    p_gs[2] = (19.3 - 0.2*p_gs[0] - 2.5*p_gs[1])/4.2
    
    erros = np.abs(np.divide(p_gs - p_gs_anterior, p_gs))*100
    print(i , p_gs, np.max(erros))
    if np.max(erros) < tolerancia:
        break


#Iteração #Porções #Erro

0 [16.19 5.61 0.48] 100.0
1 [8.98 6.17 0.49] 80.26721541082958
2 [8.28 6.23 0.50] 8.461822270989032
3 [8.21 6.23 0.50] 0.8319926566837964
4 [8.20 6.23 0.50] 0.08120525048549926
5 [8.20 6.23 0.50] 0.007920137622719033
6 [8.20 6.23 0.50] 0.0007724144466977479
7 [8.20 6.23 0.50] 7.532948937214735e-05


## ***Resolvendo o Sistema utilizando Decomposição LU***

Esse método a estudante não entendeu muito bem, mas não custa nada tentar, não é mesmo?!

In [ ]:
from scipy.linalg import solve_triangular

In [ ]:
m_lu = np.array([[128, 159, 63], [2.5, 32, 0.2], [0.2, 2.5, 4.2]], dtype = float)
m_lu

array([[128.00, 159.00, 63.00],
       [2.50, 32.00, 0.20],
       [0.20, 2.50, 0.20]])

In [ ]:
def eliminacao_LU (matriz):

    L = np.identity(matriz.shape[0])
    U = matriz.copy()

    passos = matriz.shape[0]-1
    for p in range(passos):
        pivo = U[p,p]
        for i in range(p+1,U.shape[0]):
            m = U[i,p]/pivo
            U[i] = U[i] - U[p]*m
            L[i,p] = m
            
    return L,U

In [ ]:
l,u = eliminacao_LU(m_lu)
l

array([[1.00, 0.00, 0.00],
       [0.02, 1.00, 0.00],
       [0.00, 0.08, 1.00]])

In [ ]:
u

array([[128.00, 159.00, 63.00],
       [0.00, 28.89, -1.03],
       [0.00, 0.00, 0.18]])

In [ ]:
np.dot(l,u)

array([[128.00, 159.00, 63.00],
       [2.50, 32.00, 0.20],
       [0.20, 2.50, 0.20]])

In [ ]:
m_lu_inv = np.zeros(m_lu.shape)
z = np.zeros(m_lu.shape[0])
i = np.identity(m_lu.shape[0])
i

array([[1.00, 0.00, 0.00],
       [0.00, 1.00, 0.00],
       [0.00, 0.00, 1.00]])

In [ ]:
z = solve_triangular(l,i[:,0], lower = True)
m_lu_inv[:,0] = solve_triangular(u,z)
m_lu_inv

array([[0.01, 0.00, 0.00],
       [-0.00, 0.00, 0.00],
       [-0.00, 0.00, 0.00]])

In [ ]:
z = solve_triangular(l,i[:,1], lower = True)
m_lu_inv[:,1] = solve_triangular(u,z)

z = solve_triangular(l, i[:,2], lower = True)
m_lu_inv[:,2] = solve_triangular(u,z)
m_lu_inv

array([[0.01, 0.19, -2.95],
       [-0.00, 0.02, 0.20],
       [-0.00, -0.43, 5.50]])

In [ ]:
np.dot(m_lu,m_lu_inv)

array([[1.00, 0.00, 0.00],
       [-0.00, 1.00, 0.00],
       [-0.00, -0.00, 1.00]])

## *CONCLUSÃO*

Por fim, após a análise utilizando todos os métodos, foi definido que a estudante deve ingerir 8.2 porções de 100g de lasanha de frango, 6.23 porções de 100g de risoto de limão e 0.5 porções de 100g de temaki, ou seja, 820g de lasanha, 623g de risoto e 50g de temaki por dia.

Em relação aos métodos para resolução do sistema linear, a estudante preferiu o método de Gauss Seidel, pois é um método iterativo e não é necessário fazer tudo na "força-bruta", o pior método para resolução é o da Decomposição LU, pois ele é TOTALMENTE DESNECESSÁRIO. Além disso, é muito importante saber essas ferramentas numéricas, pois se aplicam ao dia a dia de cada pessoa.

### ***Observação 1: Todos os valores referentes as informações nutricionais dos alimentos são fictícios***
### ***Observação 2: Não é recomendado pesquisar dietas no Google, procure um nutricionista.***